### PR TITLE
Add free-slip, partial-slip and no-slip momentum BCs

### PR DESCRIPTION
**Pull request by dengwirda**
This PR updates the way horizontal momentum boundary conditions are computed --- highlighting that the current MPAS discretisation applies no-slip conditions, introducing new free-slip and partial-slip conditions, and cleaning-up the way cell-vertex remapping is handled across partially masked cells/duals in general.

No-slip and free-slip BCs can be expressed in terms of $\nabla \times \mathbf{u}$ on the boundaries --- no-slip conditions (the current implementation) correspond to a 'full' relative vorticity computed on boundary vertices, while free-slip conditions are imposed by setting $\nabla \times \mathbf{u} = 0$ on boundary vertices (corresponding to no induced torque when a free-stream velocity flows tangential to a wall). I've also introduced so-called 'partial-slip' conditions (see e.g. NEMO) as an interpolation between the no-slip and free-slip cases, leading to a reduced vorticity on boundaries. See e.g. Roullet and Gaillard for confirmation of how momentum BCs should be applied in a vector-invariant formulation.

A new `lateral_walls` namelist section for BC settings is introduced, where `config_wall_slip_factor` controls the amount of slip at horizontal walls (both coastlines at the surface and masked boundaries against bathymetry in general). `config_wall_slip_factor = 0.0` is a no-slip condition, `config_wall_slip_factor = 1.0` is a free-slip condition, and values between 0.0 and 1.0 correspond to partial-slip cases.

No-slip (left) vs free-slip (right) behaviour can be seen in the channel case below (developed with @hetland) in which the initially uniform zonal flow generates strong vorticity sheets at the north/south boundaries in the no-slip case, while the flow remains quiescent in the free-slip case. 

mpaso_slip_1

This PR also updates the way vorticity and layer-thickness is computed and remapped at partially masked cells and duals at boundaries, ensuring that the numerical stencils reflect the masking. In the case of computing potential vorticity $\frac{1}{h}(\nabla \times \mathbf{u} + f)$ it's necessary to remap layer thickness from cells to duals (triangles) using the intersecting cell-dual 'kite' areas. Currently, I feel we are not accounting for masked cells/duals correctly. Consider a case at a boundary in which one of the three cells surrounding a boundary vertex is masked, and where a spatially uniform flow with layer-thickness $h$ exists inside the domain. 

In the current implementation the thicknesses from the two valid cells is remapped along with a 0 value for the masked cell. The result is divided by the full area of the dual (triangle) resulting in a vertex layer-thickness of only $\frac{2}{3} h$, inconsistent with the spatially uniform interior thickness. An alternative approach implemented here is to accumulate only the unmasked area of the dual (triangle) and use this in the remapping, which will results the expected vertex thickness of $h$.

$$h_{v_{old}} = \frac{A_1 h_{c_1} + A_2 h_{c_2} + 0}{A_{tria}} = \frac{A_{tria}}{A_{tria}} (\frac{1}{3}h_{c_1} + \frac{1}{3}h_{c_2}) = \frac{2}{3}h$$

$$h_{v_{new}} = \frac{A_1 h_{c_1} + A_2 h_{c_2}}{A_1 + A_2} = \frac{A_{tria}}{A_{tria}} (\frac{\frac{1}{3}h_{c_1} + \frac{1}{3}h_{c_2}}{\frac{1}{3} + \frac{1}{3}}) = h$$

$$A_{1,2,3} = \frac{1}{3}A_{tria}$$

- Roullet, G. and Gaillard, T., 2022. A fast monotone discretization of the rotating shallow water equations. Journal of Advances in Modeling Earth Systems, 14(2), p.e2021MS002663.

- https://www.nemo-ocean.eu/doc/node58.html

### CODE REVIEW

**Comment by xylar** [comment]
@dengwirda, same questions as for #48:  Is this ready to review?  If so, who would you like to review it before it goes to E3SM?

**Comment by dengwirda** [comment]
@xylar, let me know how you'd like the 'discussion' aspect of this to proceed --- this branch is ready to run, but I'd say that additional testing is required before review.

**Comment by xylar** [comment]
@dengwirda, typically these PRs on E3SM-Ocean-Discussion only proceed if you ping specific people to indicate that you want their feedback.  You can do that by adding them as a reviewer or just by `@`ing them and saying what kind of feedback you're looking for.  Neither of those typically requires the code to be tested, though it could make things a bit clearer if you just ping people to initiate discussion and then flag them for a review when things are tested.  The review here is typically cursory -- are things far enough along that we don't think we'll have a lengthy discussion (both in time and number of posts) on E3SM?

**Comment by dengwirda** [comment]
@xylar please feel free to add extra reviewers and/or review yourself here. 
The slip BCs are hopefully not controversial, though the implications of using them in different config.'s is something to discuss.
The updates to `layerThicknessVertex` at boundaries will/may change behaviour in general.

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F [hunk 1]
```diff
@@ -627,24 +627,36 @@ subroutine ocn_relativeVorticity_circulation(relativeVorticity, circulation, nor
...
-      !$acc                  minLevelVertexTop) &
-      !$acc          private(invAreaTri1, iVertex, iEdge, k, r_tmp)
+      !$acc          present(circulation, relativeVorticity, edgesOnVertex, &
+      !$acc                  maxLevelVertexBot, dcEdge, normalVelocity, &
+      !$acc                  minLevelVertexTop, areaTriangle, edgeSignOnVertex) &
+      !$acc          private(invAreaTri1, i, iEdge, k, r_tmp, wall_slip_factor)
 #else
-      !$omp do schedule(runtime) private(invAreaTri1, i, iEdge, k, r_tmp)
+      !$omp do schedule(runtime) &
+      !$omp          private(invAreaTri1, i, iEdge, k, r_tmp, wall_slip_factor)
```
Comment (xylar) [inline components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:643]: Is `boundaryVertex` also needed in the `present()` arguments?

**Comment by xylar** [comment]
@dengwirda, your list of reviewers looks good to me.

**Comment by xylar** [comment]
@dengwirda, I suggest you make one PR to E3SM to fix `dualArea` as you have done here.  I think we just want to do that once and for all, and it will be a non-BFB PR for sure.  Then, we can do a second PR with the boundary conditions.  That one can leave no-slip as the default for now but we can test the other options and consider switching to them before the v3 code freeze.

**Comment by mark-petersen** [comment]
@dengwirda this looks like a good addition. When you compare the current code with this PR using `config_wall_slip_factor = 0.0`, are the differences reasonably small? It that case, it looks like order-of-operations differences, so should be <1e-10 for the first few time steps. If that is the case, I think we could proceed to a PR to E3SM-Project with the default set to `config_wall_slip_factor = 0.0`.

**Comment by xylar** [comment]
@mark-petersen, I could be wrong but my understanding was that, because the area for vertices was being computed with full triangles rather than from the sum of the kites that actually exist, the differences will not be small and we will need to treat this as a (hopefully) non-climate-changing PR.  That is why I was suggesting that we make the bug fix for the vertex area first and then the rest.

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F [hunk 2]
```diff
@@ -1052,25 +1064,29 @@ subroutine ocn_diagnostic_solve_vorticity(dt, &
...
             do i = 1, vertexDegree
                iCell = cellsOnVertex(i,iVertex)
                layerThicknessVertex = layerThicknessVertex &
                                     + layerThickness(k,iCell) &
                                     * kiteAreasOnVertex(i,iVertex)
+               areaDual = areaDual + cellMask(k,iCell) &
+                                    * kiteAreasOnVertex(i,iVertex)
             end do
-            layerThicknessVertex = layerThicknessVertex*invAreaTri1
+            layerThicknessVertex = layerThicknessVertex/areaDual
```
Comment (mark-petersen) [inline components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:1089]: @xylar I see what you are saying. This is the change where `invAreaTri1` includes all three kites, while `areaDual` only includes the kites from active cells. This would be answer changing, and the changes in the loop above could go into a single PR.

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F [hunk 3]
```diff
@@ -627,24 +627,36 @@ subroutine ocn_relativeVorticity_circulation(relativeVorticity, circulation, nor
...
+               r_tmp = edgeSignOnVertex(i, iVertex) * &
+                       dcEdge(iEdge) * normalVelocity(k, iEdge)
+               circulation(k, iVertex) = circulation(k, iVertex) + r_tmp
             end do
+            ! no-slip: curl(u) unchanged
+            ! free-slip: curl(u) = 0.0
+            ! partial-slip: curl(u) reduced
+            circulation(k, iVertex) = circulation(k, iVertex) * &
+               (1.0_RKIND - wall_slip_factor * boundaryVertex(k, iVertex))
+            relativeVorticity(k, iVertex) = circulation(k, iVertex) * invAreaTri1
```
Comment (mark-petersen) [inline components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: The changes in this loop are what is needed to include `wall_slip_factor`. For this section, it is true that `wall_slip_factor = 0.0` should only change the solution due to order of operations. For a second PR for just the addition of `wall_slip_factor` and these lines, it will be easier to show that only minor changes occurred.

**Comment by dengwirda** [comment]
@mark-petersen, @xylar, yes I think you're both right here --- the `wall_slip_factor` change by itself should hopefully be non-climate-changing (though perhaps not BFB due to order-of-op type differences) but the boundary vertex thickness issue will change the way vorticity is computed in general --- at all coastlines for the surface layer, and at any cells/duals where layers are maksed against bathymetry in the deep ocean.

One reason I've grouped this work together here is that implicitly both changes relate to the way momentum boundary conditions are imposed. The `wall_slip_factor` should allow users to interpolate between no-slip and free-slip conditions explicitly, but the vertex thickness change will alter the way PV is computed on boundaries in all cases. In the example I have above, it looks like we may currently be computing too small a boundary vertex thickness (due to the masked kites), which would give too large a boundary PV value. Considering that free-slip has PV=0 on boundaries, and that no-slip uses a full PV value, this may mean that some kind of more-than-no-slip condition is being used currently. What do you think?

**Comment by xylar** [comment]
@dengwirda, I see the logic in including them together.

I think it will still be simpler in terms of how to frame these changes to the broader project if we first introduce a bug fix to the area weighting of layer thickness on vertices, knowing that the changes will not be small but that it's a bug.  Then, we can still point to that PR when we introduce the other changes here but we would point out that the change in behavior after the bug fix is small as long as you are using the no-slip boundary condition.  We would further point out that the "stealth" option to have free-slip or partial slip boundary conditions would be expected to introduce changes of a similar magnitude a and in a similar direction to the previous bug fix.  If testing is sufficient at that point, we could also advocate for making something other than no-slip the default at that point.

But I think it really will be necessary to introduce these changes in those 2 stages, rather than in one PR.

**Review by mark-petersen — APPROVED** [review]
I'm happy for these changes to move to E3SM-Project. Thanks.

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F (hunk 3 above)
Comment (sbrus89) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: @dengwirda, is there a reason to exchange loop order here? It seems like the order of operations issue could be avoided by applying the `(1.0_RKIND - wall_slip_factor * boundaryVertex(k, iVertex))` factor in a separate loop `k` loop following the original `do i = 1, vertexDegree` loop.

**Comment by cbegeman** [comment]
@xylar Thank you!

I just wanted to document that @mark-petersen, myself, and Alaina Hogan are testing this branch and will also test a version without the dual mesh area changes. There is no need for anyone to make changes to this PR right now.

**Comment by xylar** [comment]
> @xylar and all re: the dual-area issue, I'd say that it's worth considering what happens throughout the water column (where land cells are present in the mesh but masked) rather than just at the surface (where cells are fully culled out).

@dengwirda, ah, of course.  That's a very good point.

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F (hunk 3 above)
Comment (mark-petersen) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: I think the `invAreaTri1` here needs a correction. The logic is similar to the areaDual below, but opposite. Currently, next to a full land boundary, `areaTriangle(iVertex)` does not include the non-existent cells. But beside a stair-step boundary below water where the neighbor exists, `areaTriangle(iVertex)` is the full area. So here `invAreaTri1` is treating coastal boundaries versus underwater boundaries in a different way. Thinking about it, it seems each active edge should accumulate an associated area with it (the sub-triangle about that edge), and we should divide by that accumulated area.

Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: I think you're suggesting exactly the same correction as what happens below with `areaDual` and that one is needed for the same reason -- handling topography.

Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: Answering my own question, no, we don't want to divide by the full triangle area (phew, because we can't!), we only want to divide by the kite areas that have valid velocities (including the zero velocities at valid boundary edges).  If we divide by the full triangle area, we're back to creating the unintended partial slip conditions that Darren talks about toward the end of https://github.com/E3SM-Ocean-Discussion/E3SM/pull/49#issue-1672937070.

Comment (cbegeman) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: @xylar @mark-petersen I agree with all of this (and thanks @dengwirda for the important note about edges below the top layer).

**Comment by cbegeman** [comment]
My preference for a path forward would be to move the `areaDual` computation to `mpas_ocn_mesh.F` and save it as a variable called something like `areaVertex` with dimensions `nVertLevels`, `nVertices`. To be super clear, `areaVertex` would be the sum of only the active cell areas.

We would use `areaVertex` for `relativeVorticity = circulation / areaVertex` and `layerThicknessVertex = layerThicknessVertex / areaVertex`

How does that sound?

I also want to confirm that we should be determining active cells on a vertex using `cellMask`. `circulation` is summed over `minLevelVertexTop : maxLevelVertexBot` so I think `cellMask` would reflect that (`maxLevelVertexBot` being the one that is inclusive of inactive neighbors).

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F (hunk 3 above)
Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: @cbegeman and @mark-petersen, I had a bit more time to think about this.  I think this case is distinct from the one below where we are interpolating layer thickness from cells to vertices.

For relative vorticity, we are adding up circulation around valid edges.  Each edge can be thought of as having an kite area associated with it that is `1/2 * dcEdge * dvEdge`.  Half of that kite overlaps with the triangle of each neighboring vertex, so the triangle of overlap has an area `1/4 * dcEdge * dvEdge`.  In this case, I think the areas of these triangles are what we should be adding up to get the area in the denominator of the curl area.  One could argue that only half the triangle should be used for a boundary edge (because half its area lies in an invalid cell), in which case we would arrive back at the same `dualArea` as used to interpolate layer thickness from cells to vertices.  But I think this would underweight the contributions of the boundary edges with zero velocity to the relative vorticity.  Instead, I think we want a *different* area here that is the sum of the full triangles associated with each valid edge around the vertex.

As a side note, some vertices have only 2 valid edges.  In this case, both edges are boundary edges with zero velocity.  So what area we or, for that matter, what boundary condition we use is a moot point because the circulation an therefore the relative vorticity is always zero at these vertices.

**Comment by xylar** [comment]
@cbegeman, sorry for waffling on this.  I made a note about about why I think the area to use in the circulation is *not* the same as `areaDual`.  If you agree with my reasoning, I am not sure it makes sense to store `areaDual` after all, since it will only be used to interpolate layer thickness to vertices and not in the denominator of the relative vorticity calculation.

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F (hunk 3 above)
Comment (cbegeman) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: @xylar Pointed out to me that my example above would only apply to the free slip case where the vorticity would be 0 on both vertices on edge. So it's probably more helpful to consider the no slip case. Need to think through it more.

Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: > Do you think it's accurate to think of the actual boundary as the one connecting the middle of each edge where normalVelocity = 0 (except when the edge is parallel to the boundary)?

I think the boundary really follows the boundary edges so it connects vertices, not edge centers.

Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: >  If so, then we would only want relativeVorticity = 0 at the vertices with all boundary edges and relativeVorticity != 0 for vertices with one non-boundary edge because those vertices "stick out" into the flow.

To me, the vertices with 2 boundary edges and one sticking out into the flow are precisely the ones that implement the choice of boundary conditions. Whether that free edge is allowed to contribute to the curl or not determines the boundary condition.

**Comment by xylar** [comment]
Thanks for staying engaged with this, @dengwirda ! We'll look forward to your paper. I think we're stuck with our jagged boundaries for the time being but what you describe sounds really promising for a future improvement!

**Comment by cbegeman** [comment]
That sounds like an interesting approach, @dengwirda. Please do share your paper when it is available!

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F (hunk 3 above)
Comment (xylar) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:659]: I totally agree that focusing too much on getting no-slip right is silly if free-slip is the way to go. Free slip seems much easier to get right.

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F [hunk 4]
```diff
@@ -627,24 +627,36 @@ subroutine ocn_relativeVorticity_circulation(relativeVorticity, circulation, nor
...
 #ifdef MPAS_OPENACC
       !$acc parallel loop &
-      !$acc          present(circulation, relativeVorticity, areaTriangle, edgesOnVertex, &
-      !$acc                  maxLevelVertexBot, dcEdge, normalVelocity, edgeSignOnVertex, &
-      !$acc                  minLevelVertexTop) &
-      !$acc          private(invAreaTri1, iVertex, iEdge, k, r_tmp)
+      !$acc          present(circulation, relativeVorticity, edgesOnVertex, &
+      !$acc                  maxLevelVertexBot, dcEdge, normalVelocity, &
+      !$acc                  minLevelVertexTop, areaTriangle, edgeSignOnVertex) &
+      !$acc          private(invAreaTri1, i, iEdge, k, r_tmp, wall_slip_factor)
```
Comment (cbegeman) [inline components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:640]:
```suggestion
      !$acc          private(invAreaTri1, i, iEdge, k, r_tmp)
```
This change was made prior to testing. At least on some compilers, including `wall_slip_factor` in `private` results in it being assigned to 0 in the loop.

**Comment by cbegeman** [comment]
### Idealized testing

**Barotropic gyre:**

- NCC wall slip = 0. No slip case: differences in streamfunction between master and wallslip = 0 are O(1e-9) after 2 years
- CC wall slip = 1. Free slip case: error near the boundary is reduced from master to wallslip = 1 but error away from the boundary is increased. This may not be a cause for concern because the analytic solution is only accurate in the limit of $\nu_2 / L_x \beta = \infty$
    * Master 
image

 -  *  Wallslip = 1
image
 
**Barotropic channel:** with z-level and trough geometry to test free-slip when deeper levels terminate against bathymetry

- NCC wall slip = 0. Circulation and thus relative vorticity remains non-zero at boundary vertices.
  * Master
image

-  * Wallslip = 0 zTop
image

-  * Wallslip = 0 zBot
image

- CC wall slip = 1.
-  * Wallslip = 1 zTop
image

-  * Wallslip = 1 zBot
image

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F (hunk 4 above)
Comment (dengwirda) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:640]: Fixes a bug of mine.

**Comment by cbegeman** [comment]
### E3SM testing

**v3.LR.piControl**, 50 years, climo 1-50
https://e3sm.atlassian.net/wiki/spaces/pd/pages/5545886223/v3.LR.piControl-wallslip1
Increased throughflow in narrow passages:
image
No significant diffs in any other fields:
image
image
image
Screenshot 2025-11-01 at 3 12 18 PM
Screenshot 2025-11-01 at 3 11 50 PM



**v3.SORRM G-case**, 20 years, climo 11-20 (free-slip conditions on left with this branch, control with alfred3 branch on right)
https://e3sm.atlassian.net/wiki/spaces/pd/pages/5545885697/20251002.v3.SORRME3r3.GMPAS-JRA1p5-DIB-PISMF.wallslip1
No significant diffs in SST:
image
Significant diffs in seafloor velocity magnitude statistics (see upper right for min, mean, max):
image
Too early to say if AMOC diffs are significant:
image
Throughflow in narrow passages increased:
image
No significant sea ice diffs:
https://portal.nersc.gov/project/m2136/bin/iice/mpas-a.cgi?url1=https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.cbegeman/BlueTip/20251002.v3.SORRME3r3.GMPAS-JRA1p5-DIB-PISMF.wallslip1/mpas_analysis/ts_0001-0020_climo_0011-0020&label1=wallslip1&url2=https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.dcomeau/E3SMv3_dev/20250918.v3.SORRME3r3.GMPAS-JRA1p5-DIB-PISMF.alfred3/mpas_analysis/ts_0001-0020_climo_0011-0020&label2=alfred3&category=sea_ice&component=sea_ice&plot=&diff=0&dconfig=cryo&
